tower-http/src/cors.rs

```diff
@@ -0,0 +1,798 @@
+//! Middleware which adds headers for [CORS][mdn].
+//!
+//! # Example
+//!
+//! ```
+//! use http::{Request, Response, Method, header};
+//! use hyper::Body;
+//! use tower::{ServiceBuilder, ServiceExt, Service};
+//! use tower_http::cors::{CorsLayer, Any};
+//! use std::convert::Infallible;
+//!
+//! async fn handle(request: Request<Body>) -> Result<Response<Body>, Infallible> {
+//!     Ok(Response::new(Body::empty()))
+//! }
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let cors = CorsLayer::new()
+//!     // allow `GET` and `POST` when accessing the resource
+//!     .allow_methods(vec![Method::GET, Method::POST])
+//!     // allow requests from any origin
+//!     .allow_origin(Any);
+//!
+//! let mut service = ServiceBuilder::new()
+//!     .layer(cors)
+//!     .service_fn(handle);
+//!
+//! let request = Request::builder()
+//!     .header(header::ORIGIN, "https://example.com")
+//!     .body(Body::empty())
+//!     .unwrap();
+//!
+//! let response = service
+//!     .ready()
+//!     .await?
+//!     .call(request)
+//!     .await?;
+//!
+//! assert_eq!(
+//!     response.headers().get(header::ACCESS_CONTROL_ALLOW_ORIGIN).unwrap(),
+//!     "*",
+//! );
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+
+use futures_core::ready;
+use http::{
+    header::{self, HeaderName, HeaderValue},
+    request::Parts,
+    Method, Request, Response, StatusCode,
+};
+use pin_project::pin_project;
+use std::{
+    fmt,
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`Cors`] which adds headers for [CORS][mdn].
+///
+/// See the [module docs](crate::cors) for an example.
+///
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+#[derive(Debug, Clone)]
+pub struct CorsLayer {
+    allow_credentials: Option<HeaderValue>,
```

Comment:
I believe more efficient layout would be `HeaderMap`  with `insert` method used to guarantee only single header.
Just clone it around when needed?

Reply:
Interesting! Wanna elaborate why you think that? I'm not very familiar with the internals of `HeaderMap` but since `HeaderValue` is ref counted I imagined this would be faster.

Reply:
So let me break down what is `HeaderValue`:
- `Bytes`;
- `bool`

Let's see what is `Bytes`:

```
pub struct Bytes {
    ptr: *const u8,
    len: usize,
    // inlined "trait object"
    data: AtomicPtr<()>,
    vtable: &'static Vtable,
}
```

Basically 4 pointers hence `size_of::<Bytes>() == 32`
Coupled with the fact that each HeaderValue is wrapped inside Option it means 40 bytes for each field (thankfully Option space optimization helps to avoid 48 bytes layout)

Well `HeaderMap` itself is bloated as hell too, but it is only `96` bytes instead of 240 bytes (6 fields by 40 bytes)

Reply:
I see. That makes sense.

I think I'm gonna leave things as they are now. We can always optimize things later. Currently I'm focused on getting the functionality in.

Reply:
Ok, once it is merged I'll make PR.
Since you have to clone layer a lot, it would be beneficial to optimize its layout